### ISSUE TITLE
Port twisted.trial._asynctest.TestCase's dependencies in twisted.trial.util to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6009]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#6009|
|Type|enhancement|
|Created|2012-09-21 17:18:03Z|
|Branch|https://github.com/twisted/twisted/tree/trialutil-py3-6009|

There are a number functions in `twisted.trial.util` that are used by `twisted.trial._asynctest.TestCase`; they should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__6009 6009
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_trialutil_py3_6009 branches/trialutil-py3-6009
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1348247883000000 1348247883000000
changetime__1349203854000000 1349203854000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35798]) Branching to 'trialutil-py3-6009'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst commented|
|-|-|

(In [35800]) move _removeSafely tests from test_runner.py into the more suitable test_util.py. Refs [#6009](https://github.com/twisted/twisted/issues/6009)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_3"></a>|@itamarst set owner to @exarkun|
|-|-|

Started build run:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/trialutil-py3-6009

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

I didn't merge `twisted.trial._utilpy3` back in in this ticket, just to make it clearer what's going on. Instead I opened a ticket, [#6019](https://github.com/twisted/twisted/issues/6019).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun set owner to @itamarst|
|-|-|

Thanks.  This is a partial review, ran out of time before finishing:

  1. Failure on doc builder
  1. `admin/_twistedpython3.py` adds an incorrect element to the ported list, _twisted.internet._utilspy3.py_
  1. `from foo import (bar\nbaz)` style of imports is allowed now, fwiw
  1. The change to `twisted/trial/test/test_reporter.py` seems quite wrong to me.  We shouldn't be weakening assertions like this.  It's a horrible test, but it tests important things.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [35815]) Hopefully address review comments 1, 2 and 4. Refs [#6009](https://github.com/twisted/twisted/issues/6009)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @exarkun|
|-|-|

I'll keep 3 in mind. New build started with fixes (if all goes well) for the other points: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/trialutil-py3-6009

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Why `runWithWarningsSuppressed = staticmethod(utils.runWithWarningsSuppressed)` in `DeferredSuppressedWarningsTests` in `test_utilspy3.py`?
  1. in `twisted/trial/utils.py`, `profiled` shouldn't be killed, it should be deleted.  Violence metaphors are anti-social.
  1. in `test_util.py`, `test_callablesCalledInOrder` ends with a comment that no longer makes any sense: _# Because returning created Deferreds makes jml happy._.  Also, the line of code following it appears useless.
  1. It would make me very happy if `test_stripFlags` were just deleted.
  1. Seems like the `defer.py` changes suggests that porting `deferredGenerator` or `inlineCallbacks` is a pre-requisite of this ticket.  I suppose it doesn't actually matter though.
  1. `twisted/trial/test/packages.py` needs `__future__` imports.  Also a copyright header and a module docstring would not go amiss.  For that matter, what does `getModules` do anyway? :/
  1.  Also need future imports:
    1. `twisted.internet._utilspy3`
    1. `twisted.internet.test.test_utilspy3`
  1. Seems like `twisted.trial._utilpy3` should not be deleted from `admin/_twistedpython3.py` in _this_ branch.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst commented|
|-|-|

(In [35822]) Address review comments. Refs [#6009](https://github.com/twisted/twisted/issues/6009)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_10"></a>|@itamarst set owner to @exarkun|
|-|-|

  1. Added an explanatory comment. staticmethod is there so it doesn't become a ... descriptor? A method requiring self, anyway.
  1. Done.
  1. Removed.
  1. YAGNI indeed.
  1. Right. But getting those tests to pass requires `TestCase`... so we're stuck with the usual "it almost certainly works" limbo until then.
  1. It should be obvious to anyone what it does, otherwise it would have had an explanatory docstring! I rewrote it to not be insane.
  1. Done.
  1. Done.

Another build: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/trialutil-py3-6009

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @itamarst|
|-|-|

> Added an explanatory comment. staticmethod is there so it doesn't become a ... descriptor? A method requiring self, anyway.

Sure, but why did you change the test from using `utils.runWithWarningsSuppressed` to `self.runWithWarningsSuppressed` in the first place?

Apart from that, looks fine (actually, I didn't re-read the whole diff, I assume when you said "Done" you meant it was done).  If there's an extremely good reason for the staticmethod thing, go ahead and merge.  Otherwise it seems like the function should just be called from the module in the usual way, and then merge.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_12"></a>|@itamarst commented|
|-|-|

That class is subclassing the TestCase class for the non-Deferred version of `runWithWarningsSuppressed`, so that they can share the same tests. Thus a staticmethod.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

The staticmethod is on `SuppressedWarningsTests` which subclasses `twisted.trial.unittest.SynchronousTestCase` directly, and which appears to have no subclasses.  Am I missing something?

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_14"></a>|@itamarst commented|
|-|-|

Remember, there's two `runWithWarningsSuppressed`. `twisted.internet.test.test_utilpy3.DeferredSuppressedWarningsTests` subclasses `SuppressedWarningTests` and overrides `runWithWarningsSuppressed` to test the other one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

Thanks.  Not sure why I couldn't find that subclass before.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_16"></a>|@itamarst set status to `closed`|
|-|-|

(In [35861]) Merge trialutil-py3-6009.

Author: itamar
Review: exarkun
Fixes: [#6009](https://github.com/twisted/twisted/issues/6009)

Port rest of trial's utilities to Python 3.
